File: core/js/src/span_parent_identifier.ts
// Mirror of core/py/src/braintrust_core/span_parent_identifier.py.

import { PARENT_ID_FIELD } from "./db_fields";
import { ParentExperimentIds, ParentProjectLogIds } from "./object";

export enum SpanParentObjectType {
  EXPERIMENT = "experiment",
  PROJECT_LOGS = "project_logs",
}

const _OBJECT_TYPE_TO_PREFIX: Record<SpanParentObjectType, string> = {
  [SpanParentObjectType.EXPERIMENT]: "ex",
  [SpanParentObjectType.PROJECT_LOGS]: "pl",
};

const _PREFIX_TO_OBJECT_TYPE = Object.fromEntries(
  Object.entries(_OBJECT_TYPE_TO_PREFIX).map(([k, v]) => [v, k])
) as Record<string, SpanParentObjectType>;

const _SEP = ":";

export type SpanParentComponentsDict = (
  | ParentExperimentIds
  | ParentProjectLogIds
) & { [PARENT_ID_FIELD]?: string };

export class SpanParentComponents {
  public objectType: SpanParentObjectType;
  public objectId: string;
  public rowId: string;

  constructor(args: {
    objectType: SpanParentObjectType;
    objectId: string;
    rowId: string;
  }) {
    this.objectType = args.objectType;
    this.objectId = args.objectId;
    this.rowId = args.rowId;

    if (!(typeof this.objectType === "string")) {
      throw new Error("objectType must be a string");
    }
    if (!(typeof this.objectId === "string")) {
      throw new Error("objectId must be a string");
    }
    if (!(typeof this.rowId === "string")) {
      throw new Error("rowId must be a string");
    }

    const objectTypePrefix = _OBJECT_TYPE_TO_PREFIX[this.objectType];
    if (objectTypePrefix.includes(_SEP)) {
      throw new Error(
        `objectType prefix ${objectTypePrefix} may not contain separator character ${_SEP}`
      );
    }
    if (this.objectId.includes(_SEP)) {
      throw new Error(
        `objectId ${this.objectId} may not contain separator character ${_SEP}`
      );
    }
  }

  public toStr(): string {
    return [
      _OBJECT_TYPE_TO_PREFIX[this.objectType],
      this.objectId,
      this.rowId,
    ].join(_SEP);
  }

  public static fromStr(s: string): SpanParentComponents {
    const items: string[] = s.split(_SEP);

    if (items.length < 3) {
      throw new Error(
        `Serialized parent components string must have at least three components. Provided string ${s} has only ${items.length}`
      );
    }

    return new SpanParentComponents({
      objectType: _PREFIX_TO_OBJECT_TYPE[items[0]],
      objectId: items[1],
      rowId: items.slice(2).join(_SEP),
    });
  }

  public asDict(): SpanParentComponentsDict {
    const out = ((): SpanParentComponentsDict => {
      switch (this.objectType) {
        case SpanParentObjectType.EXPERIMENT:
          return { experiment_id: this.objectId };
        case SpanParentObjectType.PROJECT_LOGS:
          return { project_id: this.objectId, log_id: "g" };
        default:
          throw new Error("Impossible");
      }
    })();
    if (this.rowId) {
      out[PARENT_ID_FIELD] = this.rowId;
    }
    return out;
  }
}

File: core/py/src/braintrust_core/span_parent_identifier.py
# Serialization format for (object_type, object_id, row_id), which can be used
# uniquely identify a particular row as the parent of a span. Meant to be passed
# around as an opaque string.

import dataclasses
from enum import Enum

from .db_fields import PARENT_ID_FIELD


class SpanParentObjectType(Enum):
    EXPERIMENT = "experiment"
    PROJECT_LOGS = "project_logs"

    def __str__(self):
        return self.value


_OBJECT_TYPE_TO_PREFIX = {
    SpanParentObjectType.EXPERIMENT: "ex",
    SpanParentObjectType.PROJECT_LOGS: "pl",
}

_PREFIX_TO_OBJECT_TYPE = {v: k for k, v in _OBJECT_TYPE_TO_PREFIX.items()}

_SEP = ":"


@dataclasses.dataclass
class SpanParentComponents:
    object_type: SpanParentObjectType
    object_id: str
    # It is valid to describe just the parent object by setting an empty string
    # for `row_id`.
    row_id: str

    def __post_init__(self):
        assert isinstance(self.object_type, SpanParentObjectType)
        assert isinstance(self.object_id, str)
        assert isinstance(self.row_id, str)

        object_type_prefix = _OBJECT_TYPE_TO_PREFIX[self.object_type]
        assert _SEP not in object_type_prefix, object_type_prefix
        assert _SEP not in self.object_id, self.object_id

    def to_str(self):
        return _SEP.join([_OBJECT_TYPE_TO_PREFIX[self.object_type], self.object_id, self.row_id])

    @staticmethod
    def from_str(s):
        items = s.split(_SEP)

        if len(items) < 3:
            raise Exception(
                f"Serialized parent components string must have at least three components. Provided string {s} has has only {len(items)}"
            )

        # We cannot guarantee there is no separator character in the
        # user-controllable `row_id`, but since we can guarantee there is no
        # separator character in the other fields, we can still safely determine
        # the bounds of `row_id` in the serialized representation.
        return SpanParentComponents(
            object_type=_PREFIX_TO_OBJECT_TYPE[items[0]], object_id=items[1], row_id=_SEP.join(items[2:])
        )

    def as_dict(self):
        if self.object_type == SpanParentObjectType.EXPERIMENT:
            out = dict(experiment_id=self.object_id)
        elif self.object_type == SpanParentObjectType.PROJECT_LOGS:
            out = dict(project_id=self.object_id, log_id="g")
        if self.row_id:
            out[PARENT_ID_FIELD] = self.row_id
        return out
